handler/rfc8628/auth_handler.go

```diff
@@ -0,0 +1,76 @@
+// Copyright © 2024 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package rfc8628
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ory/fosite"
+	"github.com/ory/x/errorsx"
+)
+
+// MaxAttempts for retrying the generation of user codes.
+const MaxAttempts = 3
+
+// DeviceAuthHandler is a response handler for the Device Authorisation Grant as
+// defined in https://tools.ietf.org/html/rfc8628#section-3.1
+type DeviceAuthHandler struct {
+	Storage  RFC8628CoreStorage
+	Strategy RFC8628CodeStrategy
+	Config   interface {
+		fosite.DeviceProvider
+		fosite.DeviceAndUserCodeLifespanProvider
+	}
+}
+
+// HandleDeviceEndpointRequest implements https://tools.ietf.org/html/rfc8628#section-3.1
+func (d *DeviceAuthHandler) HandleDeviceEndpointRequest(ctx context.Context, dar fosite.DeviceRequester, resp fosite.DeviceResponder) error {
+	var err error
+
+	deviceCode, userCode, err := d.handleDeviceAuthSession(ctx, dar)
+	if err != nil {
+		return err
+	}
+
+	// Populate the response fields
+	resp.SetDeviceCode(deviceCode)
+	resp.SetUserCode(userCode)
+	resp.SetVerificationURI(d.Config.GetDeviceVerificationURL(ctx))
+	resp.SetVerificationURIComplete(d.Config.GetDeviceVerificationURL(ctx) + "?user_code=" + userCode)
+	resp.SetExpiresIn(int64(time.Until(dar.GetSession().GetExpiresAt(fosite.UserCode)).Seconds()))
+	resp.SetInterval(int(d.Config.GetDeviceAuthTokenPollingInterval(ctx).Seconds()))
+	return nil
+}
+
+func (d *DeviceAuthHandler) handleDeviceAuthSession(ctx context.Context, dar fosite.DeviceRequester) (string, string, error) {
+	var userCode, userCodeSignature string
+
+	deviceCode, deviceCodeSignature, err := d.Strategy.GenerateDeviceCode(ctx)
+	if err != nil {
+		return "", "", errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	dar.GetSession().SetExpiresAt(fosite.UserCode, time.Now().UTC().Add(d.Config.GetDeviceAndUserCodeLifespan(ctx)).Round(time.Second))
+	// Note: the retries are added here because we need to ensure uniqueness of user codes.
+	// The chances of duplicates should however be diminishing, because they are the same
+	// chance an attacker will be able to hit a valid code with few guesses. However, as
+	// used codes will probably still be around for some time before they get cleaned,
+	// the chances of hitting a duplicate here can be higher.
+	// Three retries should be plenty, as otherwise the entropy is definitely off.
```

Comment:
Understand where this is coming from, but not super happy with the fact that we need it.